### PR TITLE
Prevent FaIR errors with generalised emissions files

### PR DESCRIPTION
**Pull request by chrisroadmap**
See https://github.com/openscm/openscm-runner/issues/32

- [x] Tests added
- [x] Description in ``CHANGELOG.rst`` added (single line such as: ``(`#XX <https://github.com/openscm/openscm-runner/pull/XX>`_) Added feature which does something``)

@rgieseke does this do the job?

### CODE REVIEW

**Comment by chrisroadmap** [comment]
If there's a scenario you have as a test case, I can add it to the tests to make sure this works.

**Comment by rgieseke** [comment]
Checking it out right now, thanks for the quick reply! (And sorry for not providing a reproducible case right away...)

**Comment by rgieseke** [comment]
I've tried adding a failing test and test scenario in #34. In the RCMIP Phase 1 input data, there are multiple examples of summarizing emissions (e.g. for N2O as well), so only checking for CO2 is not enough.

**Comment by chrisroadmap** [comment]
> (e.g. for N2O as well)

FaIR does want to use the aggregated N2O and ignore the AFOLU / FF+I individual totals. Same as for other species e.g. the aerosols with multiple sectors.

**Comment by chrisroadmap** [comment]
How should I/we attack this @rgieseke ? Rebase #34 onto this and work from here?

**Comment by rgieseke** [comment]
> How should I/we attack this @rgieseke ? Rebase #34 onto this and work from here?

As you like (sorry for causing more confusion by opening another PR).

**Comment by rgieseke** [comment]
> FaIR does want to use the aggregated N2O and ignore the AFOLU / FF+I individual totals. Same as for other species e.g. the aerosols with multiple sectors.

Yes, for a dataset with multiple aggregated emissions (like RCMIP input) only checking for `|CO2` is not enough.

Without the fix here:
```
pint.errors.DimensionalityError: Cannot convert from 'CO2 * megametric_ton / a' ([carbon] * [mass] / [time]) to 'megatCO / a' ([carbon_monoxide] * [mass] / [time])
```

With the check for `|CO2`:
```
pint.errors.DimensionalityError: Cannot convert from 'CO * megametric_ton / a' ([carbon_monoxide] * [mass] / [time]) to 'gigatC / a' ([carbon] * [mass] / [time])
```

**Comment by chrisroadmap** [comment]
I've got no idea what's going on here. Why is it trying to convert CO units?

**Comment by chrisroadmap** [comment]
Think I got it

**Comment by rgieseke** [comment]
Looks good to me, thanks for looking into it!

**Comment by rgieseke** [comment]
Can you rebase that on master @chrisroadmap?

I think then we can merge it @znicholls

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 1]
```diff
@@ -214,6 +214,35 @@ def scmdf_to_emissions(
 
     for var_df in scmrun.groupby("variable"):
         variable = var_df.get_unique_meta("variable", no_duplicates=True)
+        # for scenarios providing both aggregate fossil and AFOLU CO2, ignore the total
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:217]:
```suggestion
        # for scenarios providing total CO2, ignore the total
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 2]
```diff
@@ -214,6 +214,35 @@ def scmdf_to_emissions(
...
+        if variable.endswith("|CO2"):
+            continue
+        # we don't care about the split for most emitted species
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:220]:
```suggestion
        # we don't care about the split below the AFOLU and Fossil levels
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 3]
```diff
@@ -214,6 +214,35 @@ def scmdf_to_emissions(
...
+        if variable.endswith(
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:221]:
```suggestion
        if "MAGICC AFOLU|" in variable or "MAGICC Fossil and Industrial|" in variable
```

should do the same thing and means you only need one line

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 4]
```diff
@@ -214,6 +214,35 @@ def scmdf_to_emissions(
...
+                "|MAGICC Fossil and Industrial|Industrial Sector",
+                "|MAGICC Fossil and Industrial|International Shipping",
+                "|MAGICC Fossil and Industrial|Residential Commercial Other",
+                "|MAGICC Fossil and Industrial|Solvents Production and Application",
+                "|MAGICC Fossil and Industrial|Transportation Sector",
+                "|MAGICC Fossil and Industrial|Waste",
+            )
+        ):
+            continue
+        # except we do want fossil and AFOLU separate for CO2
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:239]:
```suggestion
        # we do want fossil and AFOLU separate for CO2
```

File: tests/unit/test_fair1x_utils.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from openscm_runner.adapters.fair_adapter._scmdf_to_emissions import scmdf_to_emissions
+
+
+# All emissions that are not covered by FaIR (see
+# EMISSIONS_SPECIES_UNITS_CONTEXT) should be ignored, e.g.
+# 'Emissions|CO2' (which is summarizing 'MAGICC Fossil
```
Comment (znicholls) [inline tests/unit/test_fair1x_utils.py:6]:
```suggestion
# 'Emissions|CO2' (which is the total CO2 emissions
```

line below can be deleted too

File: tests/unit/test_fair1x_utils.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+# and Industrial' and 'MAGICC AFOLU')
+def test_emissions_to_ignore(test_scenario_ssp370_world):
+    emissions = scmdf_to_emissions(test_scenario_ssp370_world)
+
+    assert emissions.shape[1] == 40
```
Comment (znicholls) [inline tests/unit/test_fair1x_utils.py:11]: worth checking any CO2 AFOLU and CO2 Fossil values?

**Review by znicholls — CHANGES_REQUESTED** [review]

**Comment by rgieseke** [comment]
I'm wondering whether checking against the "allowed" values could be easier and future-proof against renaming scenario variables.

I'll give this a try @chrisroadmap

**Comment by znicholls** [comment]
Close in favour of #37?

**Pull request closed without merging**